dali/kernels/reduce/reduce_gpu_impl.cuh

```diff
@@ -0,0 +1,1092 @@
+    if (remaining_stages == 0)
+      return 1;
+    int log2 = ilog2(to_reduce);
+    int pow = log2 / (remaining_stages + 1);  // floor division - better to over-reduce...
+    // ...unless this would result in the following stages being no-ops.
+    // To consider: do some magic to use result of this stage directly and avoid small
+    // reductions or copying.
+    // For now, let's just avoid no-ops:
+    if (pow < remaining_stages && to_reduce >= (1 << remaining_stages))
+      pow = 1;
+    int macroblocks = 1 << pow;
+    return macroblocks;
+  }
+
+  /**
+   * @brief Initializes a full reduction - per-sample or whole batch.
+   */
+  void InitReduceAll() {
+    const int nsamples = in_shape_.num_samples();
+    int64_t max_reduced_extent = 0;
+    for (int i = 0; i < nsamples; i++) {
+      int64_t extent = volume(in_shape_.tensor_shape_span(i));
+      if (extent > max_reduced_extent) {
+        max_reduced_extent = extent;
+      }
+    }
+
+    if (max_reduced_extent == 0) {
+      stages_.resize(1);
+      ReductionStage &stage = stages_.front();
+      stage.index = 0;
+      // stage 0 is never ReduceAll, because that would require contiguous input
+      stage.kind = reduce_batch_ ? ReductionKind::All : ReductionKind::Sample;
+      stage.shape.resize(stage.kind == ReductionKind::All ? 1 : nsamples);
+      for (auto &rs : stage.shape) {
+        rs.outer = 1;
+        rs.inner = 1;
+        rs.reduced_out = 1;
+        rs.reduced_in = 0;
+      }
+      stage.is_last = true;
+      return;
+    }
+
+    int log2max = ilog2(max_reduced_extent * (reduce_batch_ ? nsamples : 1));
+    int substages = div_ceil(log2max, 15);
+    if (substages == 1 && reduce_batch_ && nsamples > 1) {
+      // cannot reduce multiple samples in one stage due to contiguity requirement
+      substages++;
+    }
+
+    stages_.resize(substages);
+
+    for (int s = 0; s < substages; s++) {
+      ReductionStage &stage = stages_[s];
+      stage.index = s;
+      // stage 0 is never ReduceAll, because that would require contiguous input
+      stage.kind = s == 0 ? ReductionKind::Sample
+                          : reduce_batch_ ? ReductionKind::All : ReductionKind::Block;
+      stage.shape.resize(stage.kind == ReductionKind::All ? 1 : nsamples);
+    }
+
+    for (int i = 0; i < nsamples; i++) {
+      int64_t r = volume(in_shape_.tensor_shape_span(i));
+      auto &rs = stages_[0].shape[i];
+      rs.reduced_in = r;
+      rs.inner = rs.outer = 1;
+    }
+
+    for (int s = 0; s < substages; s++) {
+      int remaining_stages = substages - 1 - s;
+      ReductionStage &stage = stages_[s];
+      if (stage.kind == ReductionKind::All) {
+        // ReduceAll cannot happen at stage 0
+        assert(s > 0 || nsamples == 1);
+        int64_t r = 0;
+        if (s > 0) {
+          for (auto &rs : stages_[s-1].shape)
+            r += rs.reduced_out;
+        } else {
+          r = max_reduced_extent;
```

Comment:
if ReduceAll cannot happen at `s == 0`, and we even assert this, the whole `if (s > 0) {} else {}` checks are superfluous, also the `r = max_reduced_extent` seems a bit weird.

Reply:
Stale comment - look at the assert again.